File: Arrays/Longest_subarray_with_sum_k/better_sol.cpp
#include <bits/stdc++.h>
using namespace std;

int longestSubarrayWithSumk(vector<int> a, long long k){
    map<long,long> preSumMap;
    long long sum =0;
    int maxLen = 0;
    for(int i=0; i<a.size(); i++){
        sum += a[i];
        if(sum == k){
            maxLen = max(maxLen, i+1);
        }
        long long rem = sum -k;
        if(preSumMap.find(rem) != preSumMap.end()){
            int len = i - preSumMap[rem];
            maxLen = max(maxLen, len);
        }

        if(preSumMap.find(sum) == preSumMap.end()){ // unique el in hashmap
            preSumMap[sum] = i;
        }
    }
    return maxLen;
}

int main(){
    vector<int> a = { -1, 1, 1 };
    int k = 1;

    
    int len = longestSubarrayWithSumk(a, k);

    cout << "The length of the longest subarray is: " << len << "\n";
    return 0;
}

File: Arrays/Longest_subarray_with_sum_k/better_sol.py
def get_longest_subarray(a,k):
    n = len(a)
    pre_sum_map  = {}
    sum_so_far = 0
    max_len = 0

    for i in range(n):
        sum_so_far += a[i]

        if sum_so_far == k:
            max_len = max(max_len,i+1)
        
        rem = sum_so_far - k
        if rem in pre_sum_map:
            length = i - pre_sum_map[rem]
            max_len = max(max_len,length)

        if sum_so_far not in pre_sum_map:
            pre_sum_map[sum_so_far] = i

    return max_len

a = [-1, 1, 1]
k = 1
length = get_longest_subarray(a, k)
print("The length of the longest subarray is:", length)